backends/xsmm/ceed-xsmm-tensor.c

```diff
@@ -73,22 +73,33 @@ static int CeedTensorContractApply_Xsmm(CeedTensorContract contract, CeedInt A,
   CeedInt blksize = 8, ind, nelem;
   CeedTensorContract_Xsmm *impl;
   ierr = CeedTensorContractGetData(contract, (void *)&impl); CeedChk(ierr);
-
-  // Get nelem and current dim
-  CeedScalar currdim = log(C/blksize) / log(J);
-  if (!(C % blksize) && currdim - (int)currdim < 1e-15) {
-    nelem = blksize;
-  } else {
-    nelem = 1;
-    currdim = log(C) / log(J);
-  }
+  const CeedInt dim = impl->dim;
 
   // Get kernel index
-  if (impl->tensorbasis)
+  if (impl->tensorbasis) {
+    // Get nelem and current dim
+    CeedScalar currdim = 0;
+    CeedInt isThisC = blksize;
+    while (isThisC < C && currdim < dim) {
+      currdim++;
+      isThisC *= J;
+    }
+    if (isThisC == C) {
+      nelem = blksize;
+    } else {
+      nelem = 1;
+      isThisC = nelem;
+      currdim = 0;
+      while (isThisC < C && currdim < dim) {
+        currdim++;
+        isThisC *= J;
+      }
+    }
```

Comment:
At a minimum, we need a comment to describe what this code is doing. But as you described it in the previous version (effectively 2 equations in 3 unknowns), it seems like this has non-unique solutions for at least some values of the inputs, and that means we should be more explicit instead of getting too clever here.

Reply:
Context: The kernels need to be identified by `C`, `J`, `B`, `tmode`, and `add`. I convert `C` and `J` to `nelem` and `currdim` to make it easier to create an index of kernels. So long as the calculation of `nelem` and `currdim` maps to the same 5 values above, its ok if we find a different solution to those equations.

Reply:
This seems like a mild maintenance liability that would vanish if we used a hash table or even just a sorted list of triplets.

Reply:
I've never made a hash table before, but I figured there had to be a better way to do this. Wasn't sure if there would be any performance cost to a sorted list of triples vs a few calculations.

Reply:
We use khash.h in PETSc. It's fast, header-only, and used internally by a number of other packages. https://github.com/attractivechaos/klib/blob/master/khash.h

Reply:
Any preference between khash or a sorted list?

Reply:
I added hkash. I've got some bizarre behavior from xsmm where I'm freeing an already freed kernel that I need to fix though.